### PR TITLE
Subscription events

### PR DESCRIPTION
**Pull request by mallen5311**
Heya @michael-erasmus :D
Would you maybe be up for reviewing these changes to the buda entities and services for subscription and payment events?  Happy to chat through any questions or concerns :D

### CODE REVIEW

File: buda/entities/payment_terms_and_schedule.proto [hunk 1]
```diff
@@ -0,0 +1,19 @@
+syntax = "proto3";
+
+package buda.entities;
```
Comment (michael-erasmus) [inline buda/entities/payment_terms_and_schedule.proto]: Sorry, this is something I should have given you a heads up on @mallen5311, but we might need to change the package names. 

In order for the Node.js and Python packages to work with the protobuf files, we had to have all the Messages and Services share a single namespace, so we're just using 

```protobuf
package buda;
```

File: buda/entities/status.proto [hunk 2]
```diff
@@ -0,0 +1,12 @@
+package buda.entities;
+
+message Status {
```
Comment (mallen5311) [inline buda/entities/status.proto]: would love to chat with you on this one, just to make sure I have full context 😄

File: buda/entities/payment.proto [hunk 3]
```diff
@@ -24,5 +24,5 @@ message Payment {
     PaymentType payment_type = 8;
     string payment_amount = 9;
     string payment_currency = 10;
-    int32 subscription_payment_sequence = 11;
+    PaymentTermsAndSchedule payment_terms_and_schedule = 11;
```
Comment (mallen5311) [inline buda/entities/payment.proto]: ah yeah, good one!  I meant to actually remove this one, as I think it maybe isn't the best fit in `payment` anymore

File: buda/entities/subscription_created.proto [hunk 4]
```diff
@@ -4,14 +4,20 @@ package buda.entities;
...
 import "buda/entities/payment.proto";
 
 message SubscriptionCreated {
     Uuid id = 1;
     Uuid user_id = 2;
-    google.protobuf.Timestamp recorded_at = 3;
-    Subscription subscription = 4;
-    Payment payment = 5;
-
+    Uuid plan_id = 3;
```
Comment (mallen5311) [inline buda/entities/subscription_created.proto]: Hmm... maybe we can chat on this one 😄 

I feels like maybe that SubscriptionCreated would be the only place that the additional nested Subscription message would apply.  If that is the case (and I'm not overlooking something 😅 ), wouldn't it perhaps be cleaner to not have the messages nested?

File: buda/entities/subscription_period_updated.proto [hunk 5]
```diff
@@ -0,0 +1,19 @@
+syntax = "proto3";
+
+package buda.entities;
+
+import "buda/entities/uuid.proto";
+import "google/protobuf/timestamp.proto";
+import "buda/entities/subscription.proto";
+import "buda/entities/payment.proto";
+
+message UpdatedSubscriptionPeriod {
```
Comment (mallen5311) [inline buda/entities/subscription_period_updated.proto]: ah, good catch, meant to fix this, on the same page 😄

File: buda/entities/payment_terms_and_schedule.proto (hunk 1 above)
Comment (mallen5311) [inline buda/entities/payment_terms_and_schedule.proto]: 👍

File: buda/entities/subscription_created.proto [hunk 6]
```diff
@@ -2,16 +2,30 @@ syntax = "proto3";
...
+      google.protobuf.Timestamp created_at = 3;
+      Uuid plan_id = 4;
+      string plan_name = 5;
+      string gateway_customer_id = 6;
+      PaymentTerms payment_terms = 7;
+      PaymnetSchedule payment_schedule = 8;
+      float term_value = 9;
+      google.protobuf.Timestamp initial_period_start = 10;
+      google.protobuf.Timestamp initial_period_end = 11;
+    }
```
Comment (mallen5311) [inline buda/entities/subscription_created.proto]: How does this feel, @michael-erasmus ?

File: buda/entities/payment_terms.proto [hunk 7]
```diff
@@ -0,0 +1,9 @@
+syntax = "proto3";
+
+package buda;
+
+enum PaymentTerms {
+  DUE_UPON_RECEIPT = 0;
+  NET_30 = 1;
+  NET_60 = 2;
+}
```
Comment (mallen5311) [inline buda/entities/payment_terms.proto]: I split `payment_terms` and `payment_schedule` into two different files, plus removed the message wrapper.  Does that maybe feel a bit cleaner, @michael-erasmus ?

**Comment by mallen5311** [comment]
Heya @michael-erasmus , 👋  
Made a couple changes based on our chat earlier. Mind having a look when you get a moment?

File: buda/entities/payment_terms.proto (hunk 7 above)
Comment (michael-erasmus) [inline buda/entities/payment_terms.proto]: Yeah, I like that! 👍

File: buda/entities/subscription_created.proto [hunk 8]
```diff
@@ -6,20 +6,26 @@ import "buda/entities/uuid.proto";
...
 message SubscriptionCreated {
+    message Subscription {
+      Uuid id = 1;
+      SubscriptionStatus status = 2;
+      google.protobuf.Timestamp created_at = 3;
+      Uuid plan_id = 4;
+      string plan_name = 5;
+      string gateway_customer_id = 6;
+      PaymentTerms payment_terms = 7;
+      PaymnetSchedule payment_schedule = 8;
```
Comment (michael-erasmus) [inline buda/entities/subscription_created.proto]: Small typo here :D

File: buda/entities/subscription_created.proto [hunk 9]
```diff
@@ -6,20 +6,26 @@ import "buda/entities/uuid.proto";
...
-    Uuid subscription_id = 6;
-    SubscriptionStatus status = 7;
-    google.protobuf.Timestamp created_at = 8;
-    PaymentTermsAndSchedule payment_schedule = 9;
-    float term_value = 10;
-    google.protobuf.Timestamp initial_period_start = 11;
-    google.protobuf.Timestamp initial_period_end = 12;
-    Payment payment = 13;
+    google.protobuf.Timestamp created_at = 3;
+    Subscription subscription = 4;
```
Comment (michael-erasmus) [inline buda/entities/subscription_created.proto]: Nice and clean!

File: buda/entities/subscription_created.proto [hunk 10]
```diff
@@ -6,20 +6,26 @@ import "buda/entities/uuid.proto";
 import "google/protobuf/timestamp.proto";
 import "buda/entities/payment.proto";
 import "buda/entities/subscription_status.proto";
-import "buda/entities/payment_terms_and_schedule.proto";
+import "buda/entities/payment_terms.proto"
```
Comment (michael-erasmus) [inline buda/entities/subscription_created.proto]: Need a trailing semi-colon here 😄

**Comment by michael-erasmus** [comment]
@mallen5311 by the way, the command you can use to check the compilation is `make package-python`

**Comment by mallen5311** [comment]
ah yeah, I keep forgetting to do that! 😅   Fixed 2 typos, and now compiles!

**Comment by mallen5311** [comment]
I went ahead and merged this in, after making all feedback changes and confirming all changes compile :)

**Pull request merged**